server/swap/swap.go

```diff
@@ -554,38 +358,66 @@ func (s *Swapper) ChainsSynced(base, quote uint32) (bool, error) {
 	return quoteSynced, nil
 }
 
-func (s *Swapper) restoreState(state *State, allowPartial bool) error {
-	// State binary version check should be done when State is loaded.
+func (s *Swapper) restoreActiveSwaps(allowPartial bool) error {
+	// Load active swap data from DB.
+	swapData, err := s.storage.ActiveSwaps()
+	if err != nil {
+		return err
+	}
+	log.Infof("Loaded swap data for %d active swaps.", len(swapData))
+	if len(swapData) == 0 {
+		return nil
+	}
 
-	// Check that the assets required by State are included
+	// Check that the required assets backends are available.
 	missingAssets := make(map[uint32]bool)
-	for _, id := range state.Assets {
-		if s.coins[id] == nil {
-			if !allowPartial {
-				return fmt.Errorf("unable to find backend for asset %d in restore state", id)
-			}
-			log.Warnf("Unable to find backend for asset %d in restore state.", id)
+	checkAsset := func(id uint32) {
+		if s.coins[id] == nil && !missingAssets[id] {
+			log.Warnf("Unable to find backend for asset %d with active swaps.", id)
 			missingAssets[id] = true
 		}
 	}
+	for _, sd := range swapData {
+		checkAsset(sd.Base)
+		checkAsset(sd.Quote)
+	}
+
+	if len(missingAssets) > 0 && !allowPartial {
+		return fmt.Errorf("missing backend for asset with active swaps")
+	}
 
-	// Load the matchTrackers, calling the Contract and Redemption asset.Backend
-	// methods as needed.
+	// Load the matchTrackers, calling the Contract, CoinConftime, and
+	// Redemption asset.Backend methods as needed.
+
+	type swapStatusData struct {
+		SwapAsset       uint32 // from market schema and takerSell bool
+		RedeemAsset     uint32
+		SwapTime        int64  // {a,b}ContractTime
+		ContractCoinOut []byte // {a,b}ContractCoinID
+		ContractScript  []byte // {a,b}Contract
+		RedeemTime      int64  // {a,b}RedeemTime
+		RedeemCoinIn    []byte // {a,b}aRedeemCoinID
+		// SwapConfirmTime is not in db, use Backend.CoinConfTime(ContractCoinOut, swapConf)
+	}
 
 	translateSwapStatus := func(ss *swapStatus, ssd *swapStatusData, cpSwapCoin []byte) error {
+		ss.swapAsset, ss.redeemAsset = ssd.SwapAsset, ssd.RedeemAsset
+
 		swapCoin := ssd.ContractCoinOut
 		if len(swapCoin) > 0 {
 			assetID := ssd.SwapAsset
-			swap, err := s.coins[assetID].Backend.Contract(swapCoin, ssd.ContractScript)
+			swapAsset := s.coins[assetID]
+			swap, err := swapAsset.Backend.Contract(swapCoin, ssd.ContractScript)
 			if err != nil {
 				return fmt.Errorf("unable to find swap out coin %x for asset %d: %w", swapCoin, assetID, err)
 			}
 			ss.swap = swap
 			ss.swapTime = encode.UnixTimeMilli(ssd.SwapTime)
-		}
 
-		if ssd.SwapConfirmTime != 0 {
-			ss.swapConfirmed = encode.UnixTimeMilli(ssd.SwapConfirmTime)
+			ss.swapConfirmed, err = swapAsset.Backend.CoinConfTime(swapCoin, int64(swapAsset.SwapConf))
+			if err != nil {
+				log.Warnf("No swap confirmed time for %v: %v", swap, err)
+			}
```

Comment:
Since this probably will not work because of the policies regarding block times, I think we could just give the client the benefit here too and use the current time. Market restarts are a messy business, so the more leniency we can give them, the fewer problems we'll encounter is my guess.